src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -200,6 +201,15 @@ public override void GetObjectData(SerializationInfo info, StreamingContext cont
         /// </summary>
         public Dispatcher Dispatcher { get; }
 
+        public IList<SiloAddress> GetCompatibleSiloList(GrainId grain)
+        {
+            var typeCode = grain.GetTypeCode();
+            var compatibleSilos = GrainTypeManager.GetSupportedSilos(typeCode).Intersect(AllActiveSilos).ToList();
```

Comment:
This method has to be cheap, since it's on the warm/hot path. Currently, it is too allocation-happy.